src/compiler/scala/reflect/macros/contexts/Typers.scala

```diff
@@ -41,7 +41,14 @@ trait Typers {
 
   def inferImplicitValue(pt: Type, silent: Boolean = true, withMacrosDisabled: Boolean = false, pos: Position = enclosingPosition): Tree = {
     macroLogVerbose(s"inferring implicit value of type $pt, macros = ${!withMacrosDisabled}")
-    universe.analyzer.inferImplicit(universe.EmptyTree, pt, false, callsiteTyper.context, silent, withMacrosDisabled, pos, (pos, msg) => throw TypecheckException(pos, msg))
+    // If there is an open implicit in the context, use its tree as the one for which the implicit needs to be inserted.
+    // Otherwise, use the empty tree. This prevents false divergent implicit errors like the one in
+    // https://github.com/scala/bug/issues/10398.
+    val tree = callsiteTyper.context.openImplicits match {
+      case oi :: _ => oi.tree
+      case _ => universe.EmptyTree
+    }
+    universe.analyzer.inferImplicit(tree, pt, false, callsiteTyper.context, silent, withMacrosDisabled, pos, (pos, msg) => throw TypecheckException(pos, msg))
```

Comment:
In a standard (non-macro) implicit search, the nested searches that try to fill in implicit paramaters are made in `Typer.applyImplicitArgs` with:

```
inferImplicitFor(paramTp, tree = fun, context, reportAmbiguous = context.reportErrors)
```

`fun` is the tree with an `MethodType(implicit .. )`, ie the application ends up looking like `Apply(fun, implicitArgs)`.

As the search recurses, these trees are stacked in a `List[OpenImplicit]`, which, as you've found, is scanned with:

```
(context.openImplicits find { case OpenImplicit(info, tp, tree1) => !info.sym.isMacro && tree1.symbol == tree.symbol && dominates(pt, tp)}) match {
```

to detect if the current implicitly called method appears higher up in the stack. If so, and if the current expected type isn't reducing in complexity, we declare divergence.

`implicit` macros are given a free pass in this check.

I'm wondering whether using `openImplicits.head` is really what we want here. For one thing, this would mean different behaviour depending on whether the macro was called explicitly or implicitly.

Perhaps what we should to is special case `EmptyTree` in the divergence check.